readme.txt
=== Flickr Shortcode Importer ===
Contributors: comprock
Donate link: http://peimic.com/about-peimic/donate/
Tags: flickr,featured image,import,media library,photo
Requires at least: 3.0.0
Tested up to: 3.2.1
Stable tag: 1.4.3

Imports [flickr] & [flickrset] shortcode and Flickr-sourced A/IMG tagged media into the Media Library.

== Description ==
Imports [flickr] & [flickrset] shortcode and Flickr-sourced A/IMG tagged media into the Media Library. Furthermore, it transforms the post content [flickr] shortcodes into links containing the Media Library based image of the proper size and alignment.

The first [flickr] image found in post content is set as the post's Featured Image and removed from the post content. The remaining [flickr] shortcodes are then transfromed as image links to their attachment page.  

[flickrset] shortcode is converted to [gallery] after the Flickr set images have been added to the Media Library. If 'Set Featured Image' is checked in Options, then the first image of the [flickrset] is used as such.

Flickr-sourced A/IMG tagged media is converted into [flickr] and then imported as normal. Great for finally bringing into your control all of those media items you've been using, but now Flickr is giving you 'Image is unavaiable' for. A/IMG tag is processed before IMG to prevent unexpected results.

This plugin is handy for transitioning from plugin `wordpress-flickr-manager` to your own Media Library because you have CDN services or want to move off of third party software.

There is no restore functionality. Backup beforehand or be prepared to revert every transformed post by hand via the post revision tool.

= Options =
* Set own Flickr API key
* Posts to Import
* Skip Importing Post IDs...
* Default A Tag Class
* Default Image Alignment
* Default Image Size
* Import Flickr-based A/IMG tags
* Set Captions
* Set first [flickr] image as featured image or not
* Force Set Featured Image
* Remove first [flickr] from post content or not
* Make Nice Image Title?
* Set limit of posts to be processed

= Handled shortcode & media samples =
* [flickr id="5348222727" thumbnail="small" overlay="false" size="large" group="" align="none"]
* [flickrset id="72157626986038277" thumbnail="small" photos="" overlay="true" size="large"]
* `<a class="tt-flickr tt-flickr-Medium" title="Khan Sao Road, Bangkok, Thailand" href="http://www.flickr.com/photos/comprock/4334303694/" target="_blank"><img class="alignnone" src="http://farm3.static.flickr.com/2768/4334303694_37785d0f0d.jpg" alt="Khan Sao Road, Bangkok, Thailand" width="500" height="375" /></a>`
* `<img class="alignnone" src="http://farm3.static.flickr.com/2768/4334303694_37785d0f0d.jpg" alt="Khan Sao Road, Bangkok, Thailand" width="500" height="375" />`

= Warnings =
* Using your own Flickr API Key might be necessary. Test a single import and see the results before setting your own.
* Backup your database before importing. You can use revision to revert individual posts, but doing so in mass is a major PITA.
* It's strongly recommended to deactivate plugins like WordSocial, WP Smush.it and similar to prevent extended import times. You can always enable them and run them enmasse later.
* Flickr-sourced IMG tags will now be linked to the attachment page.
* During my own imports, a post with one [flickr] entry could take a minute. Then posts with many [flickr] entries, several Flickr-source'd A/IMG tags and [flickset] with 30 or so photos took over 10-minutes to import.
* During that importing time, it'll look like nothing is happening. The progress bar only moves after each import succeeds or fails.
* I recommend setting the limit in options to 1 and then testing your installation. That sure makes for easier recovery in case something goes wrong. If something doesn't work, report it, http://wordpress.org/extend/plugins/flickr-shortcode-importer/.
* Make sure you have enough disk space. Figure on about 1 GB per 1,000 photos given your using Scissors-continued and have a maximum image size of 1280 x 1024. If your images can be larger, then you'll probably need 1 GB per 250 photos imported.

= Thank You =
* Initial code is modeled after Viper007Bond's class based Regenerate Thumbnails plugin. The AJAX status and single auto-submission operations were a big help.
* [flickr] shortcode handling code copied from Trent Gardner's very fine Flickr Manager plugin.
* Hat's off to Alison Barret for her Settings API tutorials and class My_Theme_Options.

== Installation ==
1. Upload the `flickr-shortcode-importer` folder to the `/wp-content/plugins/` directory
1. Activate the plugin through the 'Plugins' menu in WordPress
1. Edit defaults via Settings > [flickr] Options
1. Import via Tools > [flickr] Importer

== Frequently Asked Questions ==
= Can I sponsor changes? =
Yes. Any sponsoring would be greatly welcome. Please [donate](http://peimic.com/about-peimic/donate/ "Help sponsor Flickr Shortcode Importer") and let me know what's wanted

= Help, it doesn't work =
Please leave a forum entry detailing exactly what error message you received, whether or not you attempted and succeeded with a 1 item import, a copy of post content and any other steps needed to replicate your troubles. Don't forget the WP and PHP versions.

With this information, I should be able to have an idea on how to help resolve your issues.

= Importing was working, but suddenly stopped =
Is your disk quota large enough? See Warnings in Description for help.

== Screenshots ==
1. Flickr Shortcode Importer in Plugins
2. Flickr Shortcode Importer in Tools
3. Flickr Shortcode Importer progress
4. Before Flickr Shortcode Importer for [flickr]
5. After Flickr Shortcode Importer for [flickr]
6. Flickr Shortcode Importer Options
7. Before Flickr Shortcode Importer for [flickrset]
8. After Flickr Shortcode Importer for [flickrset]
9. Before Flickr Shortcode Importer for Flickr-sourced A/IMG Tag
10. After Flickr Shortcode Importer for Flickr-sourced A/IMG Tag

== Changelog ==
= trunk =
-

= 1.4.3 =
* Add option Set own Flickr API key

= 1.4.2 =
* Add option Posts to Import
* Add option Skip Importing Post IDs...
* Add screenshots 9 & 10 for Before & After Flickr Shortcode Importer for Flickr-sourced A/IMG Tag

= 1.4.1 =
* Add A/IMG, Warning and FAQ updates readme
* Resolve http://wordpress.org/support/topic/flickr-shortcode-importer-plugin?replies=3#post-2283617

= 1.4.0 =
* Production worthy

= 1.3.5 =
* Rename ChangeLog to changelog.txt
* Convert Flickr sourced IMG to [flickr]

= 1.3.4 =
* Check for camera given photo title
* html_entity_decode photo description
* Update Options screenshot
* Put progress for a/img to shortcode conversion - moved convert_flickr_sourced_tags() into ajax_process_shortcode()
* Update readme verbiage

= 1.3.3 =
* Add option Default A Tag Class

= 1.3.2 =
* Properly call cbMkReadableStr

= 1.3.1 =
* Update plugin description

= 1.3.0 =
* Fix over zealous replacement of content for A/IMG tag to [flickr] conversion
* Use cbMkReadableStr to pretty print media filenames as a title as needed
* Update Options screenshot
* Replace preg_match_all with explode and preg_match to handle single line of many A/IMG tags
* Line break after shortcodes to ensure proper reading by WordPress
* Add option Default Image Alignment
* Add option Default Image Size
* Add warnings to readme
* Import Flickr-based A/IMG tags
* Set get_shortcode_size default to medium
* Add Estimated time required to import notice

= 1.2.0 =
* Add option Set Captions
* Add option Force Set Featured Image
* Add option Make Nice Image Title?
* Allow 2 minutes per photo import before timing out
* Add Flickr Shortcode Importer before & after screenshots for [flickrset]
* Import [flickrset] shortcode content

= 1.1.0 =
* Add Flickr Shortcode Importer Options screenshot
* Add option Remove First Flickr Shortcode
* Add option to setting Featured Image or not
* Add import page link to options page
* Add conversion limit option
* Polish up About Flickr Shortcode Importer section
* Add text domain
* Rename class.settings.php to class.options.php
* Add [flickr] Options link to [flickr] Import page
* Add icon to [flickr] Import page
* Flickr Shortcode Importer Options page added
* [flickr] Options linked from Plugins
* Update pot file
* Remove old settings file
* Use Alison Barret's class My_Theme_Options
* Add ob_settings.php options page helper for using Settings API
* TODO video import
* Backup or reversion reminders

= 1.0.1 =
* Replace duplicate lookup by guid with _flickr_src in postmeta
* RenderVideo via FlickrManager code
* Remove [flickr] lookup LIMIT

= 1.0.0 =
* Initial release for production use

= 0.1.0 =
* Initial release for production testing
* Prep for further production testing;
* Remove import LIMIT;
* Turn off duplicate handling due to being unable to process guid setting or recall;
* Remove unused helper files;
* Add screenshots;
* Set version 0.1.0;
* Correct featured image setting and first_image removal handling;
* Prevent duplicate [flickr] imports;
* Revise readme content;
* Handle featured image setting;
* Remove first [flickr] from post_content;
* Correct [flickr] replacement for attachment links
* Load phpFlickr;
* In test mode;
* Verbiage updates;
* Process shortcode via do_shortcode;
* Readme thank you

ChangeLog
------------------------------------------------------------------------
r423649 | comprock | 2011-08-15 15:45:46 +0800 (Mon, 15 Aug 2011) | 3 lines

* Update Options screenshot
* Prep 1.3.4

------------------------------------------------------------------------
r423645 | comprock | 2011-08-15 15:43:41 +0800 (Mon, 15 Aug 2011) | 3 lines

* Put progress for a/img to shortcode conversion - moved convert_flickr_sourced_tags() into ajax_process_shortcode()
* Update readme verbiage

------------------------------------------------------------------------
r423534 | comprock | 2011-08-15 07:07:28 +0800 (Mon, 15 Aug 2011) | 1 line

Update ChangeLog
------------------------------------------------------------------------
r423527 | comprock | 2011-08-15 06:50:34 +0800 (Mon, 15 Aug 2011) | 2 lines

* Add option Default A Tag Class

------------------------------------------------------------------------
r423490 | comprock | 2011-08-15 04:50:13 +0800 (Mon, 15 Aug 2011) | 1 line

Update changlog
------------------------------------------------------------------------
r423489 | comprock | 2011-08-15 04:49:15 +0800 (Mon, 15 Aug 2011) | 3 lines


* Properly call cbMkReadableStr

------------------------------------------------------------------------
r423480 | comprock | 2011-08-15 04:05:52 +0800 (Mon, 15 Aug 2011) | 1 line

Update ChangeLog
------------------------------------------------------------------------
r423478 | comprock | 2011-08-15 04:00:42 +0800 (Mon, 15 Aug 2011) | 3 lines

* Update plugin description
* Release 1.3.1

------------------------------------------------------------------------
r423468 | comprock | 2011-08-15 03:30:33 +0800 (Mon, 15 Aug 2011) | 1 line

Prep for 1.3.0
------------------------------------------------------------------------
r423467 | comprock | 2011-08-15 03:29:24 +0800 (Mon, 15 Aug 2011) | 1 line

Update Changelog
------------------------------------------------------------------------
r423462 | comprock | 2011-08-15 03:15:58 +0800 (Mon, 15 Aug 2011) | 3 lines

* Fix over zealous replacement of content for A/IMG tag to [flickr] conversion
* Use cbMkReadableStr to pretty print media filenames as a title as needed

------------------------------------------------------------------------
r423438 | comprock | 2011-08-15 02:23:56 +0800 (Mon, 15 Aug 2011) | 10 lines

* Update Options screenshot
* Replace preg_match_all with explode and preg_match to handle single line of many A/IMG tags
* Line break after shortcodes to ensure proper reading by WordPress
* Add option Default Image Alignment
* Add option Default Image Size
* Add warnings to readme
* Import Flickr-based A/IMG tags
* Set get_shortcode_size default to medium
* Add Estimated time required to import notice

------------------------------------------------------------------------
r423380 | comprock | 2011-08-14 23:21:48 +0800 (Sun, 14 Aug 2011) | 4 lines

* Add warnings to readme
* Import Flickr-based A/IMG tags
* Set get_shortcode_size default to medium

------------------------------------------------------------------------
r423330 | comprock | 2011-08-14 21:18:56 +0800 (Sun, 14 Aug 2011) | 2 lines

* Add Estimated time required to import notice

------------------------------------------------------------------------
r423314 | comprock | 2011-08-14 20:41:45 +0800 (Sun, 14 Aug 2011) | 1 line

* Release 1.2.0
------------------------------------------------------------------------
r423313 | comprock | 2011-08-14 20:40:25 +0800 (Sun, 14 Aug 2011) | 1 line

Update options screenshot
------------------------------------------------------------------------
r423312 | comprock | 2011-08-14 20:37:19 +0800 (Sun, 14 Aug 2011) | 1 line

TODO update
------------------------------------------------------------------------
r423311 | comprock | 2011-08-14 20:32:43 +0800 (Sun, 14 Aug 2011) | 1 line


------------------------------------------------------------------------
r423308 | comprock | 2011-08-14 20:29:06 +0800 (Sun, 14 Aug 2011) | 2 lines

* Add option Set Captions

------------------------------------------------------------------------
r423295 | comprock | 2011-08-14 19:48:18 +0800 (Sun, 14 Aug 2011) | 2 lines

* Add option Force Set Featured Image

------------------------------------------------------------------------
r423294 | comprock | 2011-08-14 19:40:16 +0800 (Sun, 14 Aug 2011) | 2 lines

* Add option Make Nice Image Title?

------------------------------------------------------------------------
r423278 | comprock | 2011-08-14 19:14:30 +0800 (Sun, 14 Aug 2011) | 4 lines

* Import [flickrset] shortcode content
* Allow 2 minutes per photo import before timing out

class.options.php
<?php

/**
 * Flickr Shortcode Importer settings class
 *
 * @ref http://alisothegeek.com/2011/01/wordpress-settings-api-tutorial-1/
 * @since 1.0
 */
class FSI_Settings {
	
	private $sections;
	private $checkboxes;
	private $settings;
	
	/**
	 * Construct
	 *
	 * @since 1.0
	 */
	public function __construct() {
		
		// This will keep track of the checkbox options for the validate_settings function.
		$this->checkboxes				= array();
		$this->settings					= array();
		$this->get_settings();
		
		$this->sections['general']      = __( 'General Settings' , 'flickr-shortcode-importer');
		$this->sections['api']   		= __( 'Flickr API' , 'flickr-shortcode-importer');
		$this->sections['reset']        = __( 'Reset to Defaults' , 'flickr-shortcode-importer');
		$this->sections['about']        = __( 'About Flickr Shortcode Importer' , 'flickr-shortcode-importer');
		
		add_action( 'admin_menu', array( &$this, 'add_pages' ) );
		add_action( 'admin_init', array( &$this, 'register_settings' ) );

		load_plugin_textdomain( 'flickr-shortcode-importer', false, '/flickr-shortcode-importer/languages/' );
		
		if ( ! get_option( 'fsi_options' ) )
			$this->initialize_settings();
		
	}
	
	/**
	 * Add options page
	 *
	 * @since 1.0
	 */
	public function add_pages() {
		
		$admin_page = add_options_page( __( 'Flickr Shortcode Importer Options' , 'flickr-shortcode-importer'), __( '[flickr] Options' , 'flickr-shortcode-importer'), 'manage_options', 'fsi-options', array( &$this, 'display_page' ) );
		
		add_action( 'admin_print_scripts-' . $admin_page, array( &$this, 'scripts' ) );
		add_action( 'admin_print_styles-' . $admin_page, array( &$this, 'styles' ) );
		
	}
	
	/**
	 * Create settings field
	 *
	 * @since 1.0
	 */
	public function create_setting( $args = array() ) {
		
		$defaults = array(
			'id'      => 'default_field',
			'title'   => __( 'Default Field' , 'flickr-shortcode-importer'),
			'desc'    => __( 'This is a default description.' , 'flickr-shortcode-importer'),
			'std'     => '',
			'type'    => 'text',
			'section' => 'general',
			'choices' => array(),
			'class'   => ''
		);
			
		extract( wp_parse_args( $args, $defaults ) );
		
		$field_args = array(
			'type'      => $type,
			'id'        => $id,
			'desc'      => $desc,
			'std'       => $std,
			'choices'   => $choices,
			'label_for' => $id,
			'class'     => $class
		);
		
		if ( $type == 'checkbox' )
			$this->checkboxes[] = $id;
		
		add_settings_field( $id, $title, array( $this, 'display_setting' ), 'fsi-options', $section, $field_args );
	}
	
	/**
	 * Display options page
	 *
	 * @since 1.0
	 */
	public function display_page() {
		
		echo '<div class="wrap">
	<div class="icon32" id="icon-options-general"></div>
	<h2>' . __( 'Flickr Shortcode Importer Options' , 'flickr-shortcode-importer') . '</h2>';
	
		echo '<form action="options.php" method="post">';
	
		settings_fields( 'fsi_options' );
		echo '<div class="ui-tabs">
			<ul class="ui-tabs-nav">';
		
		foreach ( $this->sections as $section_slug => $section )
			echo '<li><a href="#' . $section_slug . '">' . $section . '</a></li>';
		
		echo '</ul>';
		do_settings_sections( $_GET['page'] );
		
		echo '</div>
		<p class="submit"><input name="Submit" type="submit" class="button-primary" value="' . __( 'Save Changes' , 'flickr-shortcode-importer') . '" /></p>

		<p>When ready, <a href="'.get_admin_url().'tools.php?page=flickr-shortcode-importer">'.__('begin [flickr] importing', 'flickr-shortcode-importer').'</a>
		
	</form>';
	
	echo '<script type="text/javascript">
		jQuery(document).ready(function($) {
			var sections = [];';
			
			foreach ( $this->sections as $section_slug => $section )
				echo "sections['$section'] = '$section_slug';";
			
			echo 'var wrapped = $(".wrap h3").wrap("<div class=\"ui-tabs-panel\">");
			wrapped.each(function() {
				$(this).parent().append($(this).parent().nextUntil("div.ui-tabs-panel"));
			});
			$(".ui-tabs-panel").each(function(index) {
				$(this).attr("id", sections[$(this).children("h3").text()]);
				if (index > 0)
					$(this).addClass("ui-tabs-hide");
			});
			$(".ui-tabs").tabs({
				fx: { opacity: "toggle", duration: "fast" }
			});
			
			$("input[type=text], textarea").each(function() {
				if ($(this).val() == $(this).attr("placeholder") || $(this).val() == "")
					$(this).css("color", "#999");
			});
			
			$("input[type=text], textarea").focus(function() {
				if ($(this).val() == $(this).attr("placeholder") || $(this).val() == "") {
					$(this).val("");
					$(this).css("color", "#000");
				}
			}).blur(function() {
				if ($(this).val() == "" || $(this).val() == $(this).attr("placeholder")) {
					$(this).val($(this).attr("placeholder"));
					$(this).css("color", "#999");
				}
			});
			
			$(".wrap h3, .wrap table").show();
			
			// This will make the "warning" checkbox class really stand out when checked.
			// I use it here for the Reset checkbox.
			$(".warning").change(function() {
				if ($(this).is(":checked"))
					$(this).parent().css("background", "#c00").css("color", "#fff").css("fontWeight", "bold");
				else
					$(this).parent().css("background", "none").css("color", "inherit").css("fontWeight", "normal");
			});
			
			// Browser compatibility
			if ($.browser.mozilla) 
			         $("form").attr("autocomplete", "off");
		});
	</script>
</div>';
		
	}
	
	/**
	 * Description for section
	 *
	 * @since 1.0
	 */
	public function display_section() {
		// code
	}
	
	/**
	 * Description for About section
	 *
	 * @since 1.0
	 */
	public function display_about_section() {
		
		echo					<<<EOD
			<div style="width: 50%;">
				<p><img class="alignright size-medium" title="Michael in Red Square, Moscow, Russia" src="http://peimic.com/wp-content/uploads/2009/05/michael-cannon-red-square-300x2251.jpg" alt="Michael in Red Square, Moscow, Russia" width="300" height="225" /><a href="http://wordpress.org/extend/plugins/flickr-shortcode-importer/">Flickr Shortcode Importer</a> is by <a href="mailto:michael@peimic.com">Michael Cannon</a>.</p>
				<p>He's Peichi’s man, an adventurous water rat & a TYPO3 support guru who’s living simply, roaming about & smiling more.</p>
				<p>If you like this plugin, <a href="http://peimic.com/about-peimic/donate/">please donate</a>.</p>
EOD;
		$copyright				= '<p>Copyright %s <a href="http://peimic.com">Peimic.com.</a></p>';
		$copyright				= sprintf( $copyright, date( 'Y' ) );
		echo					<<<EOD
				$copyright
			</div>
EOD;
		
	}
	
	/**
	 * HTML output for text field
	 *
	 * @since 1.0
	 */
	public function display_setting( $args = array() ) {
		
		extract( $args );
		
		$options = get_option( 'fsi_options' );
		
		if ( ! isset( $options[$id] ) && $type != 'checkbox' )
			$options[$id] = $std;
		elseif ( ! isset( $options[$id] ) )
			$options[$id] = 0;
		
		$field_class = '';
		if ( $class != '' )
			$field_class = ' ' . $class;
		
		switch ( $type ) {
			
			case 'heading':
				echo '</td></tr><tr valign="top"><td colspan="2"><h4>' . $desc . '</h4>';
				break;
			
			case 'checkbox':
				
				echo '<input class="checkbox' . $field_class . '" type="checkbox" id="' . $id . '" name="fsi_options[' . $id . ']" value="1" ' . checked( $options[$id], 1, false ) . ' /> <label for="' . $id . '">' . $desc . '</label>';
				
				break;
			
			case 'select':
				echo '<select class="select' . $field_class . '" name="fsi_options[' . $id . ']">';
				
				foreach ( $choices as $value => $label )
					echo '<option value="' . esc_attr( $value ) . '"' . selected( $options[$id], $value, false ) . '>' . $label . '</option>';
				
				echo '</select>';
				
				if ( $desc != '' )
					echo '<br /><span class="description">' . $desc . '</span>';
				
				break;
			
			case 'radio':
				$i = 0;
				foreach ( $choices as $value => $label ) {
					echo '<input class="radio' . $field_class . '" type="radio" name="fsi_options[' . $id . ']" id="' . $id . $i . '" value="' . esc_attr( $value ) . '" ' . checked( $options[$id], $value, false ) . '> <label for="' . $id . $i . '">' . $label . '</label>';
					if ( $i < count( $options ) - 1 )
						echo '<br />';
					$i++;
				}
				
				if ( $desc != '' )
					echo '<br /><span class="description">' . $desc . '</span>';
				
				break;
			
			case 'textarea':
				echo '<textarea class="' . $field_class . '" id="' . $id . '" name="fsi_options[' . $id . ']" placeholder="' . $std . '" rows="5" cols="30">' . wp_htmledit_pre( $options[$id] ) . '</textarea>';
				
				if ( $desc != '' )
					echo '<br /><span class="description">' . $desc . '</span>';
				
				break;
			
			case 'password':
				echo '<input class="regular-text' . $field_class . '" type="password" id="' . $id . '" name="fsi_options[' . $id . ']" value="' . esc_attr( $options[$id] ) . '" />';
				
				if ( $desc != '' )
					echo '<br /><span class="description">' . $desc . '</span>';
				
				break;
			
			case 'text':
			default:
		 		echo '<input class="regular-text' . $field_class . '" type="text" id="' . $id . '" name="fsi_options[' . $id . ']" placeholder="' . $std . '" value="' . esc_attr( $options[$id] ) . '" />';
		 		
		 		if ( $desc != '' )
		 			echo '<br /><span class="description">' . $desc . '</span>';
		 		
		 		break;
		 	
		}
		
	}
	
	/**
	 * Settings and defaults
	 * 
	 * @since 1.0
	 */
	public function get_settings() {
		
		/* General Settings
		===========================================*/
		
		$this->settings['set_caption'] = array(
			'section' => 'general',
			'title'   => __( 'Set Captions' , 'flickr-shortcode-importer'),
			'desc'    => __( 'Uses media title as the caption.' , 'flickr-shortcode-importer'),
			'type'    => 'checkbox',
			'std'     => 1 // Set to 1 to be checked by default, 0 to be unchecked by default.
		);
		
		$this->settings['import_flickr_sourced_tags'] = array(
			'section' => 'general',
			'title'   => __( 'Import Flickr-sourced A/IMG tags' , 'flickr-shortcode-importer'),
			'desc'    => __( 'Converts Flickr-sourced A/IMG tags to [flickr] and then proceeds with import.' , 'flickr-shortcode-importer'),
			'type'    => 'checkbox',
			'std'     => 1 // Set to 1 to be checked by default, 0 to be unchecked by default.
		);
		
		$this->settings['set_featured_image'] = array(
			'section' => 'general',
			'title'   => __( 'Set Featured Image' , 'flickr-shortcode-importer'),
			'desc'    => __( 'Set the first [flickr] or [flickrset] image found as the Featured Image. Will not replace the current Featured Image of a post.' , 'flickr-shortcode-importer'),
			'type'    => 'checkbox',
			'std'     => 1 // Set to 1 to be checked by default, 0 to be unchecked by default.
		);
		
		$this->settings['force_set_featured_image'] = array(
			'section' => 'general',
			'title'   => __( 'Force Set Featured Image' , 'flickr-shortcode-importer'),
			'desc'    => __( 'Set the Featured Image even if one already exists for a post.', 'flickr-shortcode-importer'),
			'type'    => 'checkbox',
			'std'     => 0 // Set to 1 to be checked by default, 0 to be unchecked by default.
		);
		
		$this->settings['remove_first_flickr_shortcode'] = array(
			'section' => 'general',
			'title'   => __( 'Remove First Flickr Shortcode' , 'flickr-shortcode-importer'),
			'desc'    => __( 'Remove the first [flickr] from post content? If you use Featured Images, this will help prevent duplicate images in your post.' , 'flickr-shortcode-importer'),
			'type'    => 'checkbox',
			'std'     => 1 // Set to 1 to be checked by default, 0 to be unchecked by default.
		);
		
		$this->settings['make_nice_image_title'] = array(
			'section' => 'general',
			'title'   => __( 'Make Nice Image Title?' , 'flickr-shortcode-importer'),
			'desc'    => __( "If the image title is a filename and the image is part of a Flickr set, the Flickr set title plus a numeric suffix will be used instead." , 'flickr-shortcode-importer'),
			'type'    => 'checkbox',
			'std'     => 1 // Set to 1 to be checked by default, 0 to be unchecked by default.
		);
		
		$this->settings['default_image_alignment'] = array(
			'section' => 'general',
			'title'   => __( 'Default Image Alignment' , 'flickr-shortcode-importer'),
			'desc'    => __( 'Default alignment of image displayed in post when no alignment is found.' , 'flickr-shortcode-importer'),
			'type'    => 'select',
			'std'     => 'left',
			'choices' => array(
				'none'		=> 'None',
				'left'		=> 'Left',
				'center'	=> 'Center',
				'right'		=> 'Right',
			)
		);
		
		$this->settings['default_image_size'] = array(
			'section' => 'general',
			'title'   => __( 'Default Image Size' , 'flickr-shortcode-importer'),
			'desc'    => __( 'Default size of image displayed in post when no size is found.' , 'flickr-shortcode-importer'),
			'type'    => 'select',
			'std'     => 'medium',
			'choices' => array(
				'thumbnail'	=> 'Small',
				'medium'	=> 'Medium',
				'large'		=> 'Large',
				'full'		=> 'Original',
			)
		);
		
		$this->settings['default_a_tag_class'] = array(
			'title'   => __( 'Default A Tag Class' , 'flickr-shortcode-importer'),
			'desc'    => __( "Useful for lightbox'ing." , 'flickr-shortcode-importer'),
			'std'     => '',
			'type'    => 'text',
			'section' => 'general'
		);
		
		$this->settings['posts_to_import'] = array(
			'title'   => __( 'Posts to Import' , 'flickr-shortcode-importer'),
			'desc'    => __( "A CSV list of post ids to import, like '1,2,3'." , 'flickr-shortcode-importer'),
			'std'     => '',
			'type'    => 'text',
			'section' => 'general'
		);
		
		$this->settings['skip_importing_post_ids'] = array(
			'title'   => __( 'Skip Importing Posts' , 'flickr-shortcode-importer'),
			'desc'    => __( "A CSV list of post ids to not import, like '1,2,3'." , 'flickr-shortcode-importer'),
			'std'     => '',
			'type'    => 'text',
			'section' => 'general'
		);
		
		$this->settings['limit'] = array(
			'title'   => __( 'Import Limit' , 'flickr-shortcode-importer'),
			'desc'    => __( 'Useful for testing import on a limited amount of posts. 0 or blank means unlimited.' , 'flickr-shortcode-importer'),
			'std'     => '',
			'type'    => 'text',
			'section' => 'general'
		);
		
		$this->settings['flickr_api_key'] = array(
			'title'   => __( 'Flickr API Key' , 'flickr-shortcode-importer'),
			'desc'    => __( '<a href="http://www.flickr.com/services/api/">Flickr API Documentation</a>' , 'flickr-shortcode-importer'),
			'std'     => '9f9508c77dc554c1ee7fdc006aa1879e',
			'type'    => 'text',
			'section' => 'api'
		);
		
		$this->settings['flickr_api_secret'] = array(
			'title'   => __( 'Flickr API Secret' , 'flickr-shortcode-importer'),
			'desc'    => __( '' , 'flickr-shortcode-importer'),
			'std'     => 'e63952df7d02cc03',
			'type'    => 'text',
			'section' => 'api'
		);
		
		// Here for reference
		if ( false ) {
		$this->settings['example_text'] = array(
			'title'   => __( 'Example Text Input' , 'flickr-shortcode-importer'),
			'desc'    => __( 'This is a description for the text input.' , 'flickr-shortcode-importer'),
			'std'     => 'Default value',
			'type'    => 'text',
			'section' => 'general'
		);
		
		$this->settings['example_textarea'] = array(
			'title'   => __( 'Example Textarea Input' , 'flickr-shortcode-importer'),
			'desc'    => __( 'This is a description for the textarea input.' , 'flickr-shortcode-importer'),
			'std'     => 'Default value',
			'type'    => 'textarea',
			'section' => 'general'
		);
		
		$this->settings['example_checkbox'] = array(
			'section' => 'general',
			'title'   => __( 'Example Checkbox' , 'flickr-shortcode-importer'),
			'desc'    => __( 'This is a description for the checkbox.' , 'flickr-shortcode-importer'),
			'type'    => 'checkbox',
			'std'     => 1 // Set to 1 to be checked by default, 0 to be unchecked by default.
		);
		
		$this->settings['example_heading'] = array(
			'section' => 'general',
			'title'   => '', // Not used for headings.
			'desc'    => 'Example Heading',
			'type'    => 'heading'
		);
		
		$this->settings['example_radio'] = array(
			'section' => 'general',
			'title'   => __( 'Example Radio' , 'flickr-shortcode-importer'),
			'desc'    => __( 'This is a description for the radio buttons.' , 'flickr-shortcode-importer'),
			'type'    => 'radio',
			'std'     => '',
			'choices' => array(
				'choice1' => 'Choice 1',
				'choice2' => 'Choice 2',
				'choice3' => 'Choice 3'
			)
		);
		
		$this->settings['example_select'] = array(
			'section' => 'general',
			'title'   => __( 'Example Select' , 'flickr-shortcode-importer'),
			'desc'    => __( 'This is a description for the drop-down.' , 'flickr-shortcode-importer'),
			'type'    => 'select',
			'std'     => '',
			'choices' => array(
				'choice1' => 'Other Choice 1',
				'choice2' => 'Other Choice 2',
				'choice3' => 'Other Choice 3'
			)
		);
		}
		
		/* Appearance
		===========================================*/
		
		$this->settings['header_logo'] = array(
			'section' => 'appearance',
			'title'   => __( 'Header Logo' , 'flickr-shortcode-importer'),
			'desc'    => __( 'Enter the URL to your logo for the plugin header.' , 'flickr-shortcode-importer'),
			'type'    => 'text',
			'std'     => ''
		);
		
		$this->settings['favicon'] = array(
			'section' => 'appearance',
			'title'   => __( 'Favicon' , 'flickr-shortcode-importer'),
			'desc'    => __( 'Enter the URL to your custom favicon. It should be 16x16 pixels in size.' , 'flickr-shortcode-importer'),
			'type'    => 'text',
			'std'     => ''
		);
		
		$this->settings['custom_css'] = array(
			'title'   => __( 'Custom Styles' , 'flickr-shortcode-importer'),
			'desc'    => __( 'Enter any custom CSS here to apply it to your plugin.' , 'flickr-shortcode-importer'),
			'std'     => '',
			'type'    => 'textarea',
			'section' => 'appearance',
			'class'   => 'code'
		);
				
		/* Reset
		===========================================*/
		
		$this->settings['reset_plugin'] = array(
			'section' => 'reset',
			'title'   => __( 'Reset plugin' , 'flickr-shortcode-importer'),
			'type'    => 'checkbox',
			'std'     => 0,
			'class'   => 'warning', // Custom class for CSS
			'desc'    => __( 'Check this box and click "Save Changes" below to reset plugin options to their defaults.' , 'flickr-shortcode-importer')
		);
		
	}
	
	/**
	 * Initialize settings to their default values
	 * 
	 * @since 1.0
	 */
	public function initialize_settings() {
		
		$default_settings = array();
		foreach ( $this->settings as $id => $setting ) {
			if ( $setting['type'] != 'heading' )
				$default_settings[$id] = $setting['std'];
		}
		
		update_option( 'fsi_options', $default_settings );
		
	}
	
	/**
	* Register settings
	*
	* @since 1.0
	*/
	public function register_settings() {
		
		register_setting( 'fsi_options', 'fsi_options', array ( &$this, 'validate_settings' ) );
		
		foreach ( $this->sections as $slug => $title ) {
			if ( $slug == 'about' )
				add_settings_section( $slug, $title, array( &$this, 'display_about_section' ), 'fsi-options' );
			else
				add_settings_section( $slug, $title, array( &$this, 'display_section' ), 'fsi-options' );
		}
		
		$this->get_settings();
		
		foreach ( $this->settings as $id => $setting ) {
			$setting['id'] = $id;
			$this->create_setting( $setting );
		}
		
	}
	
	/**
	* jQuery Tabs
	*
	* @since 1.0
	*/
	public function scripts() {
		
		wp_print_scripts( 'jquery-ui-tabs' );
		
	}
	
	/**
	* Styling for the plugin options page
	*
	* @since 1.0
	*/
	public function styles() {
		
		wp_register_style( 'fsi-admin', plugins_url( 'settings.css', __FILE__ ) );
		wp_enqueue_style( 'fsi-admin' );
		
	}
	
	/**
	* Validate settings
	*
	* @since 1.0
	*/
	public function validate_settings( $input ) {
		
		// TODO validate for integer CSV
		// posts_to_import
		// skip_importing_post_ids

		if ( ! isset( $input['reset_plugin'] ) ) {
			$options = get_option( 'fsi_options' );
			
			foreach ( $this->checkboxes as $id ) {
				if ( isset( $options[$id] ) && ! isset( $input[$id] ) )
					unset( $options[$id] );
			}
			
			return $input;
		}

		return false;
		
	}
	
}

$FSI_Settings					= new FSI_Settings();

function fsi_options( $option ) {
	$options					= get_option( 'fsi_options' );
	if ( isset( $options[$option] ) )
		return $options[$option];
	else
		return false;
}
?>